File: app/src/main/java/com/digicoffer/lauditor/Relationships/Adapter/IndividualDropdownAdapter.java
package com.digicoffer.lauditor.Relationships.Adapter;

import static com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils.maskEmail;
import static com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils.maskPhoneNumber;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.digicoffer.lauditor.Relationships.Model.IndividualModel;
import com.digicoffer.lauditor.R;

import java.util.ArrayList;
import java.util.List;

public class IndividualDropdownAdapter extends ArrayAdapter<IndividualModel> {

    private final LayoutInflater inflater;
    private List<IndividualModel> originalList;
    private List<IndividualModel> filteredList;

    public IndividualDropdownAdapter(@NonNull Context context, List<IndividualModel> list) {
        super(context, 0, list);
        inflater = LayoutInflater.from(context);
        this.originalList = new ArrayList<>(list);
        this.filteredList = new ArrayList<>(list);
    }

    @Override
    public int getCount() {
        return filteredList != null ? filteredList.size() : 0;
    }

    @Override
    public IndividualModel getItem(int position) {
        return filteredList != null ? filteredList.get(position) : null;
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        return createItemView(position, convertView, parent);
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        return createItemView(position, convertView, parent);
    }

    private View createItemView(int position, View convertView, ViewGroup parent) {
        View view = inflater.inflate(R.layout.dropdown_individual_xml, parent, false);

        IndividualModel model = getItem(position);
        if (model != null) {
            TextView tvName = view.findViewById(R.id.tv_name);
            TextView tvEmail = view.findViewById(R.id.tv_email);
            TextView tvMobile = view.findViewById(R.id.tv_mobile);

            String fullName = (model.getFirst_name() + " " + model.getLast_name()).trim();
            tvName.setText(fullName.isEmpty() ? "No Name" : fullName);
            tvEmail.setText(maskEmail(model.getEmail()));
            tvMobile.setText(maskPhoneNumber(model.getMobile()));
        }

        return view;
    }

    public void updateList(List<IndividualModel> newList) {
        this.originalList = new ArrayList<>(newList);
        this.filteredList = new ArrayList<>(newList);
        clear();
        addAll(filteredList);
        notifyDataSetChanged();
        Log.d("IndividualDropdownAdapter", "Updated list with " + filteredList.size() + " items");
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults results = new FilterResults();
                List<IndividualModel> filtered = new ArrayList<>();

                if (constraint == null || constraint.length() == 0) {
                    filtered.addAll(originalList);
                } else {
                    String filterPattern = constraint.toString().toLowerCase().trim();
                    for (IndividualModel item : originalList) {
                        String name = (item.getFirst_name() + " " + item.getLast_name()).toLowerCase();
                        String email = item.getEmail().toLowerCase();

                        if (name.contains(filterPattern) || email.contains(filterPattern)) {
                            filtered.add(item);
                        }
                    }
                }

                results.values = filtered;
                results.count = filtered.size();
                Log.d("IndividualDropdownAdapter", "Filtered: " + filtered.size() + " items for query: " + constraint);
                return results;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                filteredList = (List<IndividualModel>) results.values;
                clear();
                if (filteredList != null) {
                    addAll(filteredList);
                }
                notifyDataSetChanged();
            }
        };
    }
}

File: app/src/main/java/com/digicoffer/lauditor/Relationships/Adapter/IndividualDropdownAdapter.kt
package com.digicoffer.lauditor.Relationships.Adapter

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.TextView
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.Relationships.Model.IndividualModel
import java.util.ArrayList

class IndividualDropdownAdapter(context: Context, list: List<IndividualModel>) :
    ArrayAdapter<IndividualModel>(context, 0, list) {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private var originalList: List<IndividualModel> = ArrayList(list)
    private var filteredList: List<IndividualModel> = ArrayList(list)

    override fun getCount(): Int {
        return filteredList.size
    }

    override fun getItem(position: Int): IndividualModel? {
        return if (position in filteredList.indices) filteredList[position] else null
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return createItemView(position, convertView, parent)
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return createItemView(position, convertView, parent)
    }

    private fun createItemView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = inflater.inflate(R.layout.dropdown_individual_xml, parent, false)
        val model = getItem(position)
        if (model != null) {
            val tvName = view.findViewById<TextView>(R.id.tv_name)
            val tvEmail = view.findViewById<TextView>(R.id.tv_email)
            val tvMobile = view.findViewById<TextView>(R.id.tv_mobile)

            val firstName = model.first_name ?: ""
            val lastName = model.last_name ?: ""
            val fullName = "$firstName $lastName".trim()
            tvName.text = if (fullName.isEmpty()) "No Name" else fullName
            tvEmail.text = AndroidUtils.maskEmail(model.email ?: "")
            tvMobile.text = AndroidUtils.maskPhoneNumber(model.mobile ?: "")
        }
        return view
    }

    fun updateList(newList: List<IndividualModel>) {
        this.originalList = ArrayList(newList)
        this.filteredList = ArrayList(newList)
        clear()
        addAll(filteredList)
        notifyDataSetChanged()
        Log.d("IndividualDropdownAdapter", "Updated list with ${filteredList.size} items")
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence?): FilterResults {
                val results = FilterResults()
                val filtered: MutableList<IndividualModel> = ArrayList()

                if (constraint.isNullOrEmpty()) {
                    filtered.addAll(originalList)
                } else {
                    val filterPattern = constraint.toString().lowercase().trim()
                    for (item in originalList) {
                        val firstName = item.first_name ?: ""
                        val lastName = item.last_name ?: ""
                        val name = "$firstName $lastName".lowercase()
                        val email = (item.email ?: "").lowercase()

                        if (name.contains(filterPattern) || email.contains(filterPattern)) {
                            filtered.add(item)
                        }
                    }
                }

                results.values = filtered
                results.count = filtered.size
                Log.d("IndividualDropdownAdapter", "Filtered: ${filtered.size} items for query: $constraint")
                return results
            }

            @Suppress("UNCHECKED_CAST")
            override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
                filteredList = (results?.values as? List<IndividualModel>) ?: ArrayList()
                clear()
                addAll(filteredList)
                notifyDataSetChanged()
            }
        }
    }
}
